open-craft/edx-certificates
# Implement delete_certificate method

This PR introduces the following:

1. Implement the `delete_certificate` method of the `CertificateGen` class to delete certificate files from S3 and the local filesystem
2. Add a script `delete_pdfs.py` for easy testing on local development

For the merge branch, I targeted `opencraft/opencraft-release/ginkgo.1`, but the changes can also be applied cleanly to `upstream/master`.

# Certificate deletion process flow

Certificate deletion is (or will be) initiated from LMS via the [`XQueueCertInterface.del_cert`](https://github.com/edx/edx-platform/blob/nil/lms/djangoapps/certificates/queue.py#L165-L180) method, which sends a 'remove' task to the certificate `xqueue` queue. Finally, [`certificate_agent.py` will consume the task](https://github.com/open-craft/edx-certificates/blob/john2x/BB-742-delete-certificates-opencraft-release/certificate_agent.py#L91-L95) and delete the certificates.

# Manual testing

0. Check out this branch
1. Make sure you have set the S3 access id, key, and bucket values in `settings.py`
2. Create certificate pdfs with
   ```
   $ python create_pdfs.py --course-id 'course-v1:edX+DemoX+Demo_Course' --name 'Guido'

   PdfReadWarning: Xref table not zero-indexed. ID numbers for objects will not be corrected. [pdf.py:1503]
   2019-01-23 10:18:38,420 ERROR 3615 [gnupg] _util.py:172 - Error sending data: Broken pipe
   2019-01-23 10:18:39,416 INFO 3615 [certificates.gen_cert] gen_cert.py:372 - uploaded /var/tmp/generated_certs/downloads/3fc29261812342e693a9d0a49aa73816/Certificate.pdf to downloads/3fc29261812342e693a9d0a49aa73816/Certificate.pdf
   2019-01-23 10:18:39,423 INFO 3615 [certificates.gen_cert] gen_cert.py:383 - published /var/tmp/generated_certs/downloads/3fc29261812342e693a9d0a49aa73816/Certificate.pdf to /var/tmp/downloads/3fc29261812342e693a9d0a49aa73816/Certificate.pdf
   2019-01-23 10:18:39,538 INFO 3615 [certificates.gen_cert] gen_cert.py:372 - uploaded /var/tmp/generated_certs/cert/dc98749a321b482e82b5f15a653ad31a/valid.html to cert/dc98749a321b482e82b5f15a653ad31a/valid.html
   2019-01-23 10:18:39,541 INFO 3615 [certificates.gen_cert] gen_cert.py:383 - published /var/tmp/generated_certs/cert/dc98749a321b482e82b5f15a653ad31a/valid.html to /var/tmp/cert/dc98749a321b482e82b5f15a653ad31a/valid.html
   2019-01-23 10:18:39,620 INFO 3615 [certificates.gen_cert] gen_cert.py:372 - uploaded /var/tmp/generated_certs/cert/dc98749a321b482e82b5f15a653ad31a/Certificate.pdf.sig to cert/dc98749a321b482e82b5f15a653ad31a/Certificate.pdf.sig
   2019-01-23 10:18:39,622 INFO 3615 [certificates.gen_cert] gen_cert.py:383 - published /var/tmp/generated_certs/cert/dc98749a321b482e82b5f15a653ad31a/Certificate.pdf.sig to /var/tmp/cert/dc98749a321b482e82b5f15a653ad31a/Certificate.pdf.sig
   2019-01-23 10:18:39,707 INFO 3615 [certificates.gen_cert] gen_cert.py:372 - uploaded /var/tmp/generated_certs/cert/dc98749a321b482e82b5f15a653ad31a/verify.html to cert/dc98749a321b482e82b5f15a653ad31a/verify.html
   2019-01-23 10:18:39,711 INFO 3615 [certificates.gen_cert] gen_cert.py:383 - published /var/tmp/generated_certs/cert/dc98749a321b482e82b5f15a653ad31a/verify.html to /var/tmp/cert/dc98749a321b482e82b5f15a653ad31a/verify.html
   ```
   Take not of the download and verify uuid's. `3fc29261812342e693a9d0a49aa73816` and `dc98749a321b482e82b5f15a653ad31a` respectively in this case.
3. Delete the certificates
   ```
   $ python delete_pdfs.py -vdc98749a321b482e82b5f15a653ad31a -d3fc29261812342e693a9d0a49aa73816
   PdfReadWarning: Xref table not zero-indexed. ID numbers for objects will not be corrected. [pdf.py:1503]
   2019-01-23 10:19:15,931 INFO 4141 [certificates.gen_cert] gen_cert.py:297 - deleted S3 file <Key: opencraft-verify-test-john2x-com,downloads/3fc29261812342e693a9d0a49aa73816/Certificate.pdf>
   2019-01-23 10:19:15,931 INFO 4141 [certificates.gen_cert] gen_cert.py:305 - deleted web root directory /var/tmp/downloads/3fc29261812342e693a9d0a49aa73816
   2019-01-23 10:19:16,098 INFO 4141 [certificates.gen_cert] gen_cert.py:297 - deleted S3 file <Key: opencraft-verify-test-john2x-com,cert/dc98749a321b482e82b5f15a653ad31a/Certificate.pdf.sig>
   2019-01-23 10:19:16,173 INFO 4141 [certificates.gen_cert] gen_cert.py:297 - deleted S3 file <Key: opencraft-verify-test-john2x-com,cert/dc98749a321b482e82b5f15a653ad31a/valid.html>
   2019-01-23 10:19:16,255 INFO 4141 [certificates.gen_cert] gen_cert.py:297 - deleted S3 file <Key: opencraft-verify-test-john2x-com,cert/dc98749a321b482e82b5f15a653ad31a/verify.html>
   2019-01-23 10:19:16,259 INFO 4141 [certificates.gen_cert] gen_cert.py:305 - deleted web root directory /var/tmp/cert/dc98749a321b482e82b5f15a653ad31a
   ```
4. Verify the files are deleted from S3 and local webroot directory

# To Do

- [ ] changes to `edx-platform` to send the 'remove' task to `xqueue`
